src/main/webapp/components/results/transfer-list.js

```diff
@@ -0,0 +1,207 @@
+            {title}
+            {filtered.length !== columns.length
+              ? ` (${filtered.length}/${columns.length} visible)`
+              : null}
+          </Typography>
+          <TextField
+            fullWidth
+            value={text}
+            onChange={e => setText(e.target.value)}
+            placeholder="type to filter..."
+          />
+        </CardContent>
+        <List dense={true} style={{ flex: '1', overflow: 'auto' }}>
+          {filtered.filter(column => column.hidden).map(column => {
+            return (
+              <ListItem
+                key={column.title}
+                button
+                onClick={() => {
+                  column.hidden = false
+                  onSelect(column)
+                }}
+              >
+                <ListItemText primary={column.title} />
+              </ListItem>
+            )
+          })}
+        </List>
+      </Card>
+    </div>
+  )
+}
+
+const TransferList = props => {
+  const { columns, setColumnOrder } = props
+  const [state, setState] = useState(columns)
+
+  const onColumnChange = updatedColumns => {
+    setState(updatedColumns)
+    setColumnOrder(updatedColumns)
+  }
+
+  const onSelect = columns => updatedColumn => {
+    const updatedColumns = columns.map(column => {
+      if (column.title === updatedColumn.title) {
+        column.hidden = updatedColumn.hidden
+      }
+      return column
+    })
+
+    onColumnChange(updatedColumns)
+  }
+
+  return (
+    <div
+      style={{
+        display: 'flex',
+        flexDirection: 'column',
+        height: '100%',
+        maxHeight: '100%',
+        boxSizing: 'border-box',
+        overflow: 'hidden',
+      }}
+    >
+      <div
+        style={{
+          display: 'flex',
+          justifyContent: 'space-between',
+          alignItems: 'center',
+          padding: 3,
+          marginBottom: 10,
+        }}
+      >
+        <div>
+          <FormLabel required={props.required} error={props.error}>
+            {props.label}
+          </FormLabel>
+          <FormHelperText error={props.error}>
+            {props.helperText}
+          </FormHelperText>
+        </div>
+      </div>
+
+      <div
+        style={{ display: 'flex', flex: '1', overflow: 'hidden', padding: 3 }}
+      >
+        <DisplayedColumns
+          title="Displayed Columns"
+          onSelect={onSelect(state)}
+          columns={state}
+          onColumnChange={onColumnChange}
+        />
+        <div style={{ width: 20 }} />
+        <Attributes
+          title="Hidden Columns"
+          columns={columns}
+          onSelect={onSelect(state)}
+        />
+      </div>
+    </div>
+  )
+}
+
+// a little function to help us with reordering the result
+const reorder = (list, startIndex, endIndex) => {
+  const [lr] = list.splice(startIndex, 1)
+  list.splice(endIndex, 0, lr)
+  list.forEach((column, index) => (column.tableData.columnOrder = index))
```

Comment:
It's typically bad practice to modify react state arrays in place

Reply:
Changed to return a new list instead of modifying the supplied list.